### ISSUE TITLE
[LoopVectorizer] Miscompile due to incorrect tail computation

### ISSUE THREAD

**Issue by max-quazan** [llvm/llvm-project#55167]
I hope I can explain this correctly, my original repro is in Java and I can only provide a reduced test which I think is still showing the bug.

Run `opt -loop-vectorize -S` on the following test:

```llvm
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128-ni:1-p2:32:8:8:32-ni:2"
target triple = "x86_64-unknown-linux-gnu"

define void @test() #0 {
bb:
  br label %bb1

bb1:                                              ; preds = %bb3, %bb
  %tmp = phi i32 [ 6, %bb ], [ %tmp6, %bb3 ]
  %tmp2 = phi i32 [ 35902, %bb ], [ %tmp5, %bb3 ]
  br i1 true, label %bb3, label %bb11

bb3:                                              ; preds = %bb13, %bb11, %bb1
  %tmp4 = phi i32 [ %tmp, %bb1 ], [ 9, %bb13 ], [ 9, %bb11 ]
  %tmp5 = phi i32 [ %tmp2, %bb1 ], [ %tmp16, %bb13 ], [ %tmp12, %bb11 ]
  %tmp6 = add nuw nsw i32 %tmp, 1
  %tmp7 = icmp ult i32 %tmp, 181
  br i1 %tmp7, label %bb1, label %bb8

bb8:                                              ; preds = %bb3
  %tmp9 = phi i32 [ %tmp4, %bb3 ]
  %tmp10 = phi i32 [ %tmp5, %bb3 ]
  store atomic i32 %tmp9, i32 addrspace(1)* undef seq_cst, align 4
  unreachable

bb11:                                             ; preds = %bb1
  %tmp12 = add i32 %tmp2, undef
  br i1 undef, label %bb3, label %bb13

bb13:                                             ; preds = %bb11
  %tmp14 = add i32 %tmp12, undef
  %tmp15 = xor i32 undef, 1
  %tmp16 = add i32 %tmp14, %tmp15
  br label %bb3
}

attributes #0 = { "target-features"="+sse4.2" }

```

Some explanation: we end up storing value `%tmp9` which is effectively equal to `%tmp4` and `%tmp`. `%tmp` is a loop counter which starts from 6 and goes up to 181, so its value on the last iteration in the original loop is 181. And here is what vectorizer does:

```llvm
; ModuleID = './reduced.ll'
source_filename = "/home/mkazantsev/work/sandbox/azlinux/jdk17/x86_64/fastdebug/lib/server/boilerplate.ll"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128-ni:1-p2:32:8:8:32-ni:2"
target triple = "x86_64-unknown-linux-gnu"

define void @test() #0 {
bb:
  br i1 false, label %scalar.ph, label %vector.ph

vector.ph:                                        ; preds = %bb
  br label %vector.body

vector.body:                                      ; preds = %vector.body, %vector.ph
  %index = phi i32 [ 0, %vector.ph ], [ %index.next, %vector.body ]
  %vec.phi = phi <4 x i32> [ <i32 35902, i32 0, i32 0, i32 0>, %vector.ph ], [ %predphi9, %vector.body ]
  %vec.phi1 = phi <4 x i32> [ zeroinitializer, %vector.ph ], [ %predphi10, %vector.body ]
  %offset.idx = add i32 6, %index
  %0 = add i32 %offset.idx, 0
  %broadcast.splatinsert = insertelement <4 x i32> poison, i32 %0, i32 0
  %broadcast.splat = shufflevector <4 x i32> %broadcast.splatinsert, <4 x i32> poison, <4 x i32> zeroinitializer
  %1 = add i32 %offset.idx, 4
  %broadcast.splatinsert2 = insertelement <4 x i32> poison, i32 %1, i32 0
  %broadcast.splat3 = shufflevector <4 x i32> %broadcast.splatinsert2, <4 x i32> poison, <4 x i32> zeroinitializer
  %2 = add <4 x i32> %vec.phi, undef
  %3 = add <4 x i32> %vec.phi1, undef
  %4 = add <4 x i32> %2, undef
  %5 = add <4 x i32> %3, undef
  %6 = add <4 x i32> %4, undef
  %7 = add <4 x i32> %5, undef
  %predphi = select <4 x i1> zeroinitializer, <4 x i32> <i32 9, i32 9, i32 9, i32 9>, <4 x i32> %broadcast.splat
  %predphi4 = select <4 x i1> zeroinitializer, <4 x i32> <i32 9, i32 9, i32 9, i32 9>, <4 x i32> %broadcast.splat3
  %predphi5 = select <4 x i1> zeroinitializer, <4 x i32> <i32 9, i32 9, i32 9, i32 9>, <4 x i32> %predphi
  %predphi6 = select <4 x i1> zeroinitializer, <4 x i32> <i32 9, i32 9, i32 9, i32 9>, <4 x i32> %predphi4
  %predphi7 = select <4 x i1> zeroinitializer, <4 x i32> %6, <4 x i32> %vec.phi
  %predphi8 = select <4 x i1> zeroinitializer, <4 x i32> %7, <4 x i32> %vec.phi1
  %predphi9 = select <4 x i1> zeroinitializer, <4 x i32> %2, <4 x i32> %predphi7
  %predphi10 = select <4 x i1> zeroinitializer, <4 x i32> %3, <4 x i32> %predphi8
  %index.next = add nuw i32 %index, 8
  %8 = icmp eq i32 %index.next, 176
  br i1 %8, label %middle.block, label %vector.body, !llvm.loop !0

middle.block:                                     ; preds = %vector.body
  %bin.rdx = add <4 x i32> %predphi10, %predphi9
  %9 = call i32 @llvm.vector.reduce.add.v4i32(<4 x i32> %bin.rdx)
  %cmp.n = icmp eq i32 176, 176
  %10 = extractelement <4 x i32> %predphi6, i32 3
  br i1 %cmp.n, label %bb8, label %scalar.ph

scalar.ph:                                        ; preds = %bb, %middle.block
  %bc.resume.val = phi i32 [ 182, %middle.block ], [ 6, %bb ]
  %bc.merge.rdx = phi i32 [ 35902, %bb ], [ %9, %middle.block ]
  br label %bb1

bb1:                                              ; preds = %bb3, %scalar.ph
  %tmp = phi i32 [ %bc.resume.val, %scalar.ph ], [ %tmp6, %bb3 ]
  %tmp2 = phi i32 [ %bc.merge.rdx, %scalar.ph ], [ %tmp5, %bb3 ]
  br i1 true, label %bb3, label %bb11

bb3:                                              ; preds = %bb13, %bb11, %bb1
  %tmp4 = phi i32 [ %tmp, %bb1 ], [ 9, %bb13 ], [ 9, %bb11 ]
  %tmp5 = phi i32 [ %tmp2, %bb1 ], [ %tmp16, %bb13 ], [ %tmp12, %bb11 ]
  %tmp6 = add nuw nsw i32 %tmp, 1
  %tmp7 = icmp ult i32 %tmp, 181
  br i1 %tmp7, label %bb1, label %bb8, !llvm.loop !2

bb8:                                              ; preds = %middle.block, %bb3
  %tmp9 = phi i32 [ %tmp4, %bb3 ], [ %10, %middle.block ]
  %tmp10 = phi i32 [ %tmp5, %bb3 ], [ %9, %middle.block ]
  store atomic i32 %tmp9, i32 addrspace(1)* undef seq_cst, align 4
  unreachable

bb11:                                             ; preds = %bb1
  %tmp12 = add i32 %tmp2, undef
  br i1 undef, label %bb3, label %bb13

bb13:                                             ; preds = %bb11
  %tmp14 = add i32 %tmp12, undef
  %tmp15 = xor i32 undef, 1
  %tmp16 = add i32 %tmp14, %tmp15
  br label %bb3
}

; Function Attrs: nocallback nofree nosync nounwind readnone willreturn
declare i32 @llvm.vector.reduce.add.v4i32(<4 x i32>) #1

attributes #0 = { "target-features"="+sse4.2" }
attributes #1 = { nocallback nofree nosync nounwind readnone willreturn }

!0 = distinct !{!0, !1}
!1 = !{!"llvm.loop.isvectorized", i32 1}
!2 = distinct !{!2, !3, !1}
!3 = !{!"llvm.loop.unroll.runtime.disable"}
```

Note that vector body goes to `middle.block` which in turn has the following exit condition:

```llvm
  %cmp.n = icmp eq i32 176, 176
  %10 = extractelement <4 x i32> %predphi6, i32 3
  br i1 %cmp.n, label %bb8, label %scalar.ph
```

and then

```llvm
bb8:                                              ; preds = %middle.block, %bb3
  %tmp9 = phi i32 [ %tmp4, %bb3 ], [ %10, %middle.block ]
  %tmp10 = phi i32 [ %tmp5, %bb3 ], [ %9, %middle.block ]
  store atomic i32 %tmp9, i32 addrspace(1)* undef seq_cst, align 4
  unreachable
```

So the stored value is now `%tmp9` which is equal to `%10`. Now let's see how `%10` is computed:

```llvm
  %10 = extractelement <4 x i32> %predphi6, i32 3
  %predphi6 = select <4 x i1> zeroinitializer, <4 x i32> <i32 9, i32 9, i32 9, i32 9>, <4 x i32> %predphi4
  %predphi4 = select <4 x i1> zeroinitializer, <4 x i32> <i32 9, i32 9, i32 9, i32 9>, <4 x i32> %broadcast.splat3
```

So effectively it is `%broadcast.splat3`. Now, let's note that vector loop exits when `index.next` = 176, meaning that `index` on the last iteration is 176 - 8 = 168. Now,

```llvm
  %index = phi i32 [ 0, %vector.ph ], [ %index.next, %vector.body ] // = 168
  %offset.idx = add i32 6, %index // 174
  %1 = add i32 %offset.idx, 4 // 178
  %broadcast.splatinsert2 = insertelement <4 x i32> poison, i32 %1, i32 0
  %broadcast.splat3 = shufflevector <4 x i32> %broadcast.splatinsert2, <4 x i32> poison, <4 x i32> zeroinitializer
```

So effectively `%broadcast.splat3` is splatted 178.

So now we store 178 instead of 181.

**Comment by fhahn** [comment]
The issue is that the exit value is not modeled in VPlan and during codegen we miss that we need to generate the last value. Same issue as #51366.  Will address the outstanding comments for the fix ( https://reviews.llvm.org/D123537) tomorrow.

**Comment by max-quazan** [comment]
JFYI: this patches fixes the original miscompile in Java, so I think your assessment is correct.

**Comment by fhahn** [comment]
Should be fixed by 3bebec6592a4dd8a1fb330382fc8dd4162c6bf44.
